packages/xsnap/src/xsnap.c

```diff
@@ -0,0 +1,1350 @@
+				goto bail;
+			if (!FD_ISSET(the->connection, &fds))
+				goto bail;
+		}
+		else
+			goto bail;
+	}
+	flag = 0;
+	ioctlsocket(the->connection, FIONBIO, &flag);
+}
+#else
+{
+	int flag;
+	the->connection = socket(AF_INET, SOCK_STREAM, 0);
+	if (the->connection <= 0)
+		goto bail;
+	c_signal(SIGPIPE, SIG_IGN);
+#if mxMacOSX
+	{
+		int set = 1;
+		setsockopt(the->connection, SOL_SOCKET, SO_NOSIGPIPE, (void *)&set, sizeof(int));
+	}
+#endif
+	flag = fcntl(the->connection, F_GETFL, 0);
+	fcntl(the->connection, F_SETFL, flag | O_NONBLOCK);
+	if (connect(the->connection, (struct sockaddr*)&address, sizeof(address)) < 0) {
+		 if (errno == EINPROGRESS) {
+			fd_set fds;
+			struct timeval timeout = { 2, 0 }; // 2 seconds, 0 micro-seconds
+			int error = 0;
+			unsigned int length = sizeof(error);
+			FD_ZERO(&fds);
+			FD_SET(the->connection, &fds);
+			if (select(the->connection + 1, NULL, &fds, NULL, &timeout) == 0)
+				goto bail;
+			if (!FD_ISSET(the->connection, &fds))
+				goto bail;
+			if (getsockopt(the->connection, SOL_SOCKET, SO_ERROR, &error, &length) < 0)
+				goto bail;
+			if (error)
+				goto bail;
+		}
+		else
+			goto bail;
+	}
+	fcntl(the->connection, F_SETFL, flag);
+	c_signal(SIGPIPE, SIG_DFL);
+}
+#endif
+	return;
+bail:
+	fxDisconnect(the);
+}
+
+void fxDisconnect(txMachine* the)
+{
+#if mxWindows
+	if (the->connection != INVALID_SOCKET) {
+		closesocket(the->connection);
+		the->connection = INVALID_SOCKET;
+	}
+	WSACleanup();
+#else
+	if (the->connection >= 0) {
+		close(the->connection);
+		the->connection = -1;
+	}
+#endif
+}
+
+txBoolean fxIsConnected(txMachine* the)
+{
+	return (the->connection != mxNoSocket) ? 1 : 0;
+}
+
+txBoolean fxIsReadable(txMachine* the)
+{
+	return 0;
+}
+
+void fxReceive(txMachine* the)
+{
+	int count;
+	if (the->connection != mxNoSocket) {
+#if mxWindows
+		count = recv(the->connection, the->debugBuffer, sizeof(the->debugBuffer) - 1, 0);
+		if (count < 0)
+			fxDisconnect(the);
+		else
+			the->debugOffset = count;
+#else
+	again:
+		count = read(the->connection, the->debugBuffer, sizeof(the->debugBuffer) - 1);
+		if (count < 0) {
+			if (errno == EINTR)
+				goto again;
+			else
+				fxDisconnect(the);
+		}
+		else
+			the->debugOffset = count;
+#endif
+	}
+	the->debugBuffer[the->debugOffset] = 0;
+}
+
+void fxSend(txMachine* the, txBoolean more)
+{
+	if (the->connection != mxNoSocket) {
+#if mxWindows
+		if (send(the->connection, the->echoBuffer, the->echoOffset, 0) <= 0)
+			fxDisconnect(the);
+#else
+	again:
+		if (write(the->connection, the->echoBuffer, the->echoOffset) <= 0) {
+			if (errno == EINTR)
+				goto again;
+			else
+				fxDisconnect(the);
+		}
+#endif
+	}
+}
+
+#endif /* mxDebug */
+
+
+/**
+ * @param dest: where to store address of newly allocated buffer,
+ *		  which is null-terminated for compatibility with C strings
+ * @returns count of bytes read or -1 for error;
+ *			caller takes ownership of *dest unless an error occurs
```

Comment:
I need to update this. I changed the protocol to something more C idiomatic.

Reply:
fine. I wonder why the "resolve conversation" button isn't showing up for me.

Reply:
Let's punt on the function comment.

The norm in XS code is no comments. This is static (not exported) and the implementation is entirely straightforward. And we don't comment `fxWriteNetString`